### DISCUSSION TITLE
Is there a way to obfuscate CSS classes on production environment?

### DISCUSSION THREAD

**Discussion by juanmadurand** [vercel/next.js#13942] [Help]
Hi everyone!
I'm not using CSS Modules, just global imports of **sass** files.

With this stylesheet:
```
.sidebar {
  width: 300px;
  &_title {
    font-size: 18px;
  }
}
```

I get a minified and compiled version:
`.sidebar {width: 300px}.sidebar_title {font-size: 18px}`

Buy I'd like to get obfuscated classes, this would also help to make classNames shorter:
`.1d7e62fa{width: 300px}.fb86b8t1{font-size: 18px}`

Maybe I could add a custom plugin on `next.config.js` like this:
```
plugins: [
  [ 'css-modules-transform', {
    generateScopedName: NODE_ENV === 'development' ? '[local]' : '[hash:8]',
    extensions: ['.css', '.scss']
  }]
],
```
But I didn't find any examples actually implementing something like this on the latest Next.js stable version. Is it possible to achieve this without using CSS modules?

Thanks!

**Comment by websocket98765** [comment]
> I'm not using CSS Modules, just global imports of sass files.

Did you figure out a way to use a non-module import of SCSS at the component level? I'd be grateful if you could explain how 🙏  (I just get `SyntaxError: Unexpected token '.'`)

**Comment by IRediTOTO** [comment]
I want to know about this too. Please post here if you find a solution

**Comment by AfrazHussain** [comment]
Did someone actually manage to figure this out? Would love to have some sort of help here.

**Comment by sushilbalami** [comment]
> Hi everyone!
> I'm not using CSS Modules, just global imports of **sass** files.
> 
> With this stylesheet:
> 
> ```
> .sidebar {
>   width: 300px;
>   &_title {
>     font-size: 18px;
>   }
> }
> ```
> 
> I get a minified and compiled version:
> `.sidebar {width: 300px}.sidebar_title {font-size: 18px}`
> 
> Buy I'd like to get obfuscated classes, this would also help to make classNames shorter:
> `.1d7e62fa{width: 300px}.fb86b8t1{font-size: 18px}`
> 
> Maybe I could add a custom plugin on `next.config.js` like this:
> 
> ```
> plugins: [
>   [ 'css-modules-transform', {
>     generateScopedName: NODE_ENV === 'development' ? '[local]' : '[hash:8]',
>     extensions: ['.css', '.scss']
>   }]
> ],
> ```
> 
> But I didn't find any examples actually implementing something like this on the latest Next.js stable version. Is it possible to achieve this without using CSS modules?
> 
> Thanks!

Try

```
{ 
  test: /\.css$/,
  use: [
    {
      loader: 'style-loader'
    },
    {
      loader: 'css-loader',
      options: {
        modules: true,
        importLoaders: 1,
        localIdentName: '[sha1:hash:hex:4]'
      }
    }
  ]
}
```

**Reply by xcorman** [reply]
It doesn't work, do you have any working example?

**Reply by AfrazHussain** [reply]
I've posted the configuration that works for me as a separate answer, see if that works for you!

**Comment by AfrazHussain** [comment]
I use the following for my `next.config.js`'s webpack configuration:
```
webpack(config, { dev }) {
  if (!dev) {
    config.module.rules[2].oneOf.forEach((moduleLoader, i) => {
      Array.isArray(moduleLoader.use) &&
        moduleLoader.use.forEach((l) => {
          if (
            l.loader.includes("\\css-loader") &&
            !l.loader.includes("postcss-loader")
          ) {
            l.options = {
              ...l.options,
              modules: l.options.modules
                ? {
                    ...l.options.modules,
                    getLocalIdent: getCssModuleLocalIdent,
                  }
                : l.options.module,
            };
          }
        });
    });
  }

  return config;
}
```

where `getCssModuleLocalIdent` would be:

```
const loaderUtils = require("next/dist/compiled/loader-utils");
const path = require("path");

module.exports = (context, _, exportName, options) => {
  const relativePath = path
    .relative(context.rootContext, context.resourcePath)
    .replace(/\\+/g, "/");

  // Generate a hash to make the class name unique.
  const hash = loaderUtils.getHashDigest(
    Buffer.from(`filePath:${relativePath}#className:${exportName}`),
    "md5",
    "base64",
    5
  );

  // Have webpack interpolate the `[folder]` or `[name]` to its real value.
  return (
    loaderUtils
      .interpolateName(context, hash, options)
      .replace(
        // Webpack name interpolation returns `about.module_root__2oFM9` for
        // `.root {}` inside a file named `about.module.css`. Let's simplify
        // this.
        /\.module_/,
        "_"
      )
      // Replace invalid symbols with underscores instead of escaping
      // https://mathiasbynens.be/notes/css-escapes#identifiers-strings
      .replace(/[^a-zA-Z0-9-_]/g, "_")
      // "they cannot start with a digit, two hyphens, or a hyphen followed by a digit [sic]"
      // https://www.w3.org/TR/CSS21/syndata.html#characters
      .replace(/^(\d|--|-\d)/, "__$1")
  );
};
```


Got there after significant trial and error (and also referencing the [next.js webpack code](https://github.com/vercel/next.js/tree/canary/packages/next/bundles/webpack/packages)), but this works well with CSS module files that Next.js supports natively. 

P.S. This configuration will only work on Webpack v5. For Webpack v4, please let me know, there's a slightly different configuration that I can pull up 

Edit 1: Added `getCssModuleLocalIdent` function.

**Reply by xcorman** [reply]
does this work for SASS and SCSS?

**Reply by AfrazHussain** [reply]
I haven't tried this on SASS and SCSS, but technically it should work. All what this configuration is doing is that its forcing the `getLocalIndent` property of the `css-loader` to be a custom function that forces obfuscation.

Oh, I just realized, I forgot to include the `getCssModuleLocalIdent` function. Edited the answer with that.

**Reply by xcorman** [reply]
I am getting `getCssModuleLocalIdent is not defined` when trying to build

**Reply by AfrazHussain** [reply]
Please see the updated answer.

**Reply by xcorman** [reply]
where I have to put this `getCssModuleLocalIdent` code block?

**Reply by AfrazHussain** [reply]
You can put it in a separate file and import it in your `next.config.js` as such: `const getCssModuleLocalIdent = require("./path/to/your/file");`

**Reply by xcorman** [reply]
Dont know for some reason its not working, I am using TS with Tailwind. Tried to build but I am getting same classes.

**Reply by xcorman** [reply]
@AfrazHussain I setup a new project and now I am getting `TypeError: Cannot read property 'forEach' of undefined`. Which are the packages needed additionally?

**Reply by AfrazHussain** [reply]
Like I said in my first message, you need to make sure that you are using Webpack v5 for that to work. Next.js 11 comes with Webpack 5 enabled by default. Anything lower than that would not work. Let me know if you have v4 and I'll try pulling up the code for you.

Also this would only work in production, not in development. Doesn't make sense to obfuscate classes in development.

**Reply by philipmoniaga** [reply]
@AfrazHussain Any idea how to do it for scss ? am i need to modify sass-loader?

**Comment by IRediTOTO** [comment]
In my opinion, let's use TailwindCss.com
And use it with mangle-css-class nom package :) perfect result for me.

**Reply by xcorman** [reply]
Can you share your webpack config coz its not building at all. And are you using NextJS?

Thanks

**Reply by xcorman** [reply]
Added like this:
```
const MangleCssClassPlugin = require('mangle-css-class-webpack-plugin');

module.exports = {
  webpack: (config, options) => {
    config.plugins = [
    new MangleCssClassPlugin({
      classNameRegExp: "[cl]-[a-z][a-zA-Z0-9_]*",
      log: true,
    }),
  ];

    return config;
  }
};
```

**Reply by IRediTOTO** [reply]
yes, reply me if you can improve this code
```
webpack: (config, { dev }) => {
        config.resolve.modules.push(path.resolve('./'))

        if (!dev) {
            const MangleCssClassPlugin = require('mangle-css-class-webpack-plugin')
            config.plugins.push(
                new MangleCssClassPlugin({
                    classNameRegExp:
                        '(([a-zA-Z-:]*)[\\\\\\\\]*:)*([\\\\\\\\]*!)?tw-[a-zA-Z-]([a-zA-Z0-9-]*([\\\\\\\\]*(\\%|\\#|\\.|\\[|\\]))*)*',
                    //  ignorePrefixRegExp: "((hover|focus|active|disabled|visited|first|last|odd|even|group-hover|focus-within|xs|sm|md||lg|xl)(\\\\\\\\\\\\\\\\|\\\\)?:)*",

                    classGenerator: (original, opts, context) => {
                        if (classNames[original]) {
                            return classNames[original]
                        }

                        let nextId

                        do {
                            // Class name cannot start with a number.
                            nextId = generateClassName()
                        } while (/^[0-9_-]/.test(nextId))

                        return (classNames[original] = nextId)
                    },
                    //  log: true
                })
            )
        }
        return config
    },
```

**Reply by IRediTOTO** [reply]
There is an error when you use `tw-w-1/3` (This wont work) let use `tw-w-[50%]`

**Reply by derpdead** [reply]
Is there anybody how made obfuscating for scss? Without css modules

**Comment by n4j1Br4ch1D** [comment]
Check out this postcss plugin:
https://www.npmjs.com/package/postcss-obfuscator
